File: practica/python/8.1.py
from msvcrt import getch

#структура
DataStruct = {
    "Marka" : [], #марка автомобіля
    "PIB" : [], #піб власника
    "Year" : [], #рік випуску
    "Number" : [] #реєстраційний номер (5 символів)
}

#функція перевірки року випуску
def FuncCheckYear(Y):
    if( (Y <= 2022) and (Y >= 1900) ):
        return True

    print("\t\t\t", "Введіть рік в діапазоні [1900 2022]")
    return False

#функція перевірки реєстраційного номеру
def FuncCheckNumber(N):
    if ( not ( (N <= 99999) and (N >= 10000) ) ):
        print("\t\t\t", "Введіть рік в діапазоні [100000 99999]")
        return False
    
    for i in range(0, len(DataStruct["Number"]), 1):
        if(DataStruct["Number"][i] == N):
            print("\t\t\t", "Реєстраційний номер належить іншому користувачу")
            return False

    return True

#ГОЛОВНА ПРОГРАМА
#цикл вводу користувачів
while(True):
    print("\t", "Введіть дані користувача №", len(DataStruct["Marka"]) + 1)
    M = input("\t\t марка: ")
    P = input("\t\t піб: ")
    Y = int(input("\t\t рік випуску: "))
    N = int(input("\t\t реєстраційний номер: "))

    #перевірка на правильність введення даних і долучення користвувача
    if( FuncCheckYear(Y) and FuncCheckNumber(N) ):
        print("\t\t\t", "Користувач долучений!")
        
        DataStruct["Marka"].append(M)
        DataStruct["PIB"].append(P)
        DataStruct["Year"].append(Y)
        DataStruct["Number"].append(N)

    print("\n\n\t", "-----------", end = "\n\t")
    
    Answer = input("\t\tБажаєте долучити ще користувача? (\"+\" - так): ")

    print("\t", "-----------", end = "\n\n")
    
    #вихід з циклу
    if( (Answer != "+") or (len(DataStruct["Marka"]) == 100) ):
        if(len(DataStruct["Marka"]) >= 3):
            break

        else:
            print("\t\t", "Вихід неможливий, у вас введено менше 3 користувачів")

print("\n\n", "Вивід користувачів [рік - піб - марка - номер]:")
for i in range(0, len(DataStruct["Marka"]), 1):
    print("\t", "[", DataStruct["Year"][i], end = ", \"")
    print(DataStruct["PIB"][i], end = "\", \"")
    print(DataStruct["Marka"][i], end = "\", ")
    print(DataStruct["Number"][i], "]")

#пункт 1
SearchMarka = input("\n\n Введіть марку для пошуку кількості автомобілів цієї марки: ")

CounterSearchMarka = 0

for i in range(0, len(DataStruct["Marka"]), 1):
    if(DataStruct["Marka"][i] == SearchMarka):
        CounterSearchMarka = CounterSearchMarka + 1

print("\t", "Кількість: ", CounterSearchMarka, end = "\n\n")

#пункт 2
print("Вивід цих автомобілів зазначеної марки впорядкованим за піб і з реєстраційним номером: ")

for i in range(1, len(DataStruct["Marka"]), 1):
    for j in range(0, len(DataStruct["Marka"]) - i, 1):
        if(DataStruct["PIB"][j] > DataStruct["PIB"][j + 1]):
            TempVar = DataStruct["PIB"][j]
            DataStruct["PIB"][j] = DataStruct["PIB"][j + 1]
            DataStruct["PIB"][j + 1] = TempVar
            TempVar = DataStruct["Marka"][j]
            DataStruct["Marka"][j] = DataStruct["Marka"][j + 1]
            DataStruct["Marka"][j + 1] = TempVar
            TempVar = DataStruct["Year"][j]
            DataStruct["Year"][j] = DataStruct["Year"][j + 1]
            DataStruct["Year"][j + 1] = TempVar
            TempVar = DataStruct["Number"][j]
            DataStruct["Number"][j] = DataStruct["Number"][j + 1]
            DataStruct["Number"][j + 1] = TempVar

for i in range(0, len(DataStruct["Marka"]), 1):
    if(CounterSearchMarka > 0):
        if(DataStruct["Marka"][i] == SearchMarka):
            print("\t", "[", "\"", DataStruct["PIB"][i], "\", ", DataStruct["Number"][i], "]")
    else:
        print("\t користувачі зазначеної марки відсутні!")

        break

print(end = "\n\n")

#пункт 3
#цикл вводу
MinYear = 0
MaxYear = 0
while(True): 
    MinYear = int(input("Вивід мінімального року: "))
    MaxYear = int(input("Вивід максимального року: "))
    
    if( FuncCheckYear(MinYear) and FuncCheckYear(MaxYear) ):
        if(MaxYear >= MinYear):
            break

        else:
            print("\t\t", "Мінімальний рік більший за максимальний, перевведіть!", end = "\n\n")

print("\n", "Вивід марок і реєстраційних номерів за вказаним діапазон років:")

for i in range(0, len(DataStruct["Marka"]), 1):
    if( (DataStruct["Year"][i] <= MaxYear) and (DataStruct["Year"][i] >= MinYear) ):
        print("\t", "[", "\"", DataStruct["Marka"][i], "\", ", DataStruct["Number"][i], "]")

getch()

File: practica/pascal/8.1.pas
USES
    crt;

TYPE
    DataStruct = record
        Marka : string; //марка автомобіля
        PIB : string; //піб власника
        Year : integer; //рік випуску
        Number : integer; //реєстраційний номер (5 символів)
    end;

VAR
    MyArray : array [0..99] of DataStruct;
    CounterMyArray : integer;
    M, P : string;
    Y, N : integer;
    Answer, SearchMarka : string;
    CounterSearchMarka, MinYear, MaxYear : integer;
    TempVar : DataStruct;

FUNCTION FuncCheckYear() : boolean; //функція перевірки року випуску
begin
    if ( (Y <= 2022) and (Y >= 1900) ) then
    begin
        FuncCheckYear := true;
    end
    else
    begin
        write('            ');
        writeln('Введіть рік в діапазоні [1900; 2022]');
        writeln();
        writeln();
        
        FuncCheckYear := false;
    end;
end;

FUNCTION FuncCheckNumber() : boolean; //функція перевірки реєстраційного номеру
var
    ReturnBool : boolean;
begin
    if ( not( (N <= 99999) and (N >= 10000) ) ) then //перевірка діапазону (5 цифр)
    begin
        write('            ');
        writeln('Реєстраційний номер введений неправильному діапазоні [10000; 99999]');
        writeln();
        writeln();
        
        FuncCheckNumber := false;
    end
    else
    begin

        ReturnBool := false;

        for i : integer := 0 to (CounterMyArray - 1) do //перевірка наявності такого номеру у масиві
        begin
            if (MyArray[i].Number = N) then
            begin
                write('            ');
                writeln('Реєстраційний номер належить іншому користувачу');
                writeln();
                writeln();
                
                ReturnBool := true;

                break;
            end;
        end;

        if (ReturnBool) then
        begin
            FuncCheckNumber := false;
        end
        else
        begin
            FuncCheckNumber := true;
        end;
    end;
end;

BEGIN
    CounterMyArray := 0;

    writeln('Ввід користувачів (мінімум 3 користувачі)');

    //цикл вводу користувачів
    while (true) do
    begin
        writeln('    ', 'Введіть дані користувача №', CounterMyArray + 1);
        write('        ', 'марка: ');
        readln(M);
        write('        ', 'піб: ');
        readln(P);
        write('        ', 'рік випуску: ');
        readln(Y);
        write('        ', 'реєстраційний номер: ');
        readln(N);

        //перевірка на правильність введення даних і долучення користвувача
        if ( FuncCheckYear() and FuncCheckNumber() ) then
        begin
            writeln('            ', 'Користувач долучений!');
            
            MyArray[CounterMyArray].Marka := M;
            MyArray[CounterMyArray].PIB := P;
            MyArray[CounterMyArray].Year := Y;
            MyArray[CounterMyArray].Number := N;

            CounterMyArray := CounterMyArray + 1;
        end;

        writeln();
        writeln('    ', '-----------');
        write('        ');
        write('Бажаєте долучити ще користувача? ("+" - так): ');
        readln(Answer);
        writeln('    ', '-----------');
        writeln();
        
        //вихід з циклу
        if ( (Answer <> '+') or (CounterMyArray = 100) ) then
        begin
            if (CounterMyArray >= 3) then
            begin
                break;
            end
            else
            begin
                write('        ');
                writeln('Вихід неможливий, у вас введено менше 3 користувачів');
                writeln();
                writeln();
            end;
        end;
    end;

    writeln();
    writeln();
    writeln('Вивід користувачів [рік - піб - марка - номер]:');
    
    for i : integer := 0 to (CounterMyArray - 1) do
    begin
        write('    ', '[' , MyArray[i].Year , ', ');
        write('"', MyArray[i].PIB , '"', ', '); 
        write('"', MyArray[i].Marka , '"', ', '); 
        writeln(MyArray[i].Number , ']');
    end;

    //пункт 1
    writeln();
    writeln();
    write('Введіть марку для пошуку кількості автомобілів цієї марки: ');
    readln(SearchMarka);

    CounterSearchMarka := 0;

    for i : integer := 0 to (CounterMyArray - 1) do
    begin
        if (MyArray[i].Marka = SearchMarka) then
        begin
            CounterSearchMarka := CounterSearchMarka + 1;
        end;
    end;

    writeln('    ', 'Кількість: ', CounterSearchMarka);
    writeln();

    //пункт 2
    writeln();
    writeln();
    writeln('Вивід цих автомобілів зазначеної марки впорядкованим за піб і з реєстраційним номером: ');

    for i : integer := 1 to (CounterMyArray - 1) do
    begin
        for j : integer := 0 to (CounterMyArray - 1 - i) do
        begin
            if (MyArray[j].PIB > MyArray[j + 1].PIB) then
            begin
                TempVar := MyArray[j];
                MyArray[j] := MyArray[j + 1];
                MyArray[j + 1] := TempVar;
            end;
        end;
    end;

    for i : integer := 0 to (CounterMyArray - 1) do
    begin
        if (CounterSearchMarka > 0) then
        begin
            if (MyArray[i].Marka = SearchMarka) then
            begin
                writeln('    ', '[', '"', MyArray[i].PIB, '"', ', ' , MyArray[i].Number, ']');
            end;
        end
        else
        begin
            writeln('    ', 'користувачі зазначеної марки відсутні!');
            
            break;
        end;
    end;

    writeln();
    writeln();

    //пункт 3
    while (true) do //цикл вводу
    begin
        write('Вивід мінімального року: ');
        readln(MinYear);
        write('Вивід максимального року: ');
        readln(MaxYear);
        
        if ( FuncCheckYear() and FuncCheckYear() ) then
        begin
            if (MaxYear >= MinYear) then
            begin
                break;
            end
            else
            begin
                writeln('        ', 'Мінімальний рік більший за максимальний, перевведіть!');
                writeln();
                writeln();
            end;
        end;
    end;

    writeln();
    writeln('Вивід марок і реєстраційних номерів за вказаним діапазон років:');

    for i : integer := 0 to (CounterMyArray - 1) do
    begin
        if ( (MyArray[i].Year <= MaxYear) and (MyArray[i].Year >= MinYear) ) then
        begin
            writeln('    ', '[', '"', MyArray[i].Marka, '"', ', ' , MyArray[i].Number, ']');
        end;
    end;
    
    readkey();
END.